wal/wal_watcher.go

```diff
@@ -0,0 +1,573 @@
+	segmentCheckPeriod = 100 * time.Millisecond
+	consumer           = "consumer"
+)
+
+var (
+	watcherRecordsRead = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "records_read_total",
+			Help:      "Number of records read by the WAL watcher from the WAL.",
+		},
+		[]string{consumer, "type"},
+	)
+	watcherRecordDecodeFails = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "record_decode_failures_total",
+			Help:      "Number of records read by the WAL watcher that resulted in an error when decoding.",
+		},
+		[]string{consumer},
+	)
+	watcherSamplesSentPreTailing = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "samples_sent_pre_tailing_total",
+			Help:      "Number of sample records read by the WAL watcher and sent to remote write during replay of existing WAL.",
+		},
+		[]string{consumer},
+	)
+	watcherCurrentSegment = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "prometheus",
+			Subsystem: "wal_watcher",
+			Name:      "current_segment",
+			Help:      "Current segment the WAL watcher is reading records from.",
+		},
+		[]string{consumer},
+	)
+	lrMetrics = NewLiveReaderMetrics(prometheus.DefaultRegisterer)
+)
+
+// This function is copied from prometheus/prometheus/pkg/timestamp to avoid adding vendor to TSDB repo.
+
+// FromTime returns a new millisecond timestamp from a time.
+func FromTime(t time.Time) int64 {
+	return t.Unix()*1000 + int64(t.Nanosecond())/int64(time.Millisecond)
+}
+
+func init() {
+	prometheus.MustRegister(watcherRecordsRead)
+	prometheus.MustRegister(watcherRecordDecodeFails)
+	prometheus.MustRegister(watcherSamplesSentPreTailing)
+	prometheus.MustRegister(watcherCurrentSegment)
+}
+
+type writeTo interface {
+	Append([]record.RefSample) bool
+	StoreSeries([]record.RefSeries, int)
+	SeriesReset(int)
+}
+
+// WALWatcher watches the TSDB WAL for a given WriteTo.
+type WALWatcher struct {
+	name           string
+	writer         writeTo
+	logger         log.Logger
+	walDir         string
+	lastCheckpoint string
+	reg            prometheus.Registerer
+
+	startTime int64
+
+	recordsReadMetric       *prometheus.CounterVec
+	recordDecodeFailsMetric prometheus.Counter
+	samplesSentPreTailing   prometheus.Counter
+	currentSegmentMetric    prometheus.Gauge
+
+	quit chan struct{}
+	done chan struct{}
+
+	// For testing, stop when we hit this segment.
+	maxSegment int
+}
+
+// NewWALWatcher creates a new WAL watcher for a given WriteTo.
+func NewWALWatcher(logger log.Logger, reg prometheus.Registerer, name string, writer writeTo, walDir string) *WALWatcher {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	if reg != nil {
+		// We can't use MustRegister because WALWatcher's are recreated on config changes within Prometheus.
+		reg.Register(watcherRecordsRead)
```

Comment:
These seem to be registered both here and in `init`.

Also, should these metrics be scoped to a `WALWatcher` similar to the `LiveReader` change recently done?